iwyu.cc

```diff
@@ -3623,7 +3651,7 @@ class IwyuAstConsumer
   // Called whenever a variable, function, enum, etc is used.
   bool VisitDeclRefExpr(clang::DeclRefExpr* expr) {
     if (CanIgnoreCurrentASTNode())  return true;
-    ReportDeclUse(CurrentLoc(), expr->getDecl());
+    ReportDeclUse(CurrentLoc(), expr->getFoundDecl());
```

Comment:
This is the change that triggered the other IWYU warnings from badinc.cc to change.  getFoundDecl may or may not return the same decl as getDecl.  The details are a little vague, but it seems the getFoundDecl is sort of more specific in that it includes nested types, which is causing things like I1_TemplateMethodOnlyClass<I1_Struct, I1_Union>::t to show up whereas previously only I1_TemplateMethodOnlyClass would.  I'm not sure what the exact right thing is.  It seems to me that we probably in general want the more specific thing, but in the case of nested types, this seems superfluous because the a missing parent decl implies a missing child decl.  In the mean time, I can restore the previous functionality by changing the code as follows;

'''cpp
  bool VisitDeclRefExpr(clang::DeclRefExpr* expr) {
    if (CanIgnoreCurrentASTNode())  return true;
    if (const UsingShadowDecl* found = DynCastFrom(expr->getFoundDecl())) {
      ReportDeclUse(CurrentLoc(), found);
    } else {
      ReportDeclUse(CurrentLoc(), expr->getDecl());
    }
    return Base::VisitDeclRefExpr(expr);
  }
'''